### PR TITLE
ebdse-118 https://github.com/riptano/ebdse/issues/118

### PR DESCRIPTION
**Pull request by phact**

### CODE REVIEW

**Comment by phact** [comment]
@jshook can you review please?

File: at-stdout/src/main/java/io/engineblock/activitytypes/stdout/StdoutActivity.java [hunk 1]
```diff
@@ -132,15 +132,16 @@ protected Writer createPrintWriter() {
         List<StmtDef> stmts = stmtsDocList.getStmts(tagfilter);
         if (stmts.size() > 0) {
             for (StmtDef stmt : stmts) {
+                ParsedStmt parsed = stmt.getParsed().orError();
+
+                new StringCompositor(stmt.getStmt());
```
Comment (jshook) [inline at-stdout/src/main/java/io/engineblock/activitytypes/stdout/StdoutActivity.java]: extraneous?

File: at-stdout/src/main/java/io/engineblock/activitytypes/stdout/StdoutActivity.java [hunk 2]
```diff
@@ -132,15 +132,16 @@ protected Writer createPrintWriter() {
...
+                StringCompositor compositor = new StringCompositor(stmt.getStmt());
+                Map<String, String> bindingsMap = parsed.getBindings();
                 BindingsTemplate bt = new BindingsTemplate(stmt.getBindings());
-                String statement = stmt.getStmt();
-                Objects.requireNonNull(statement);
-                if (!statement.endsWith("\n") && getParams().getOptionalBoolean("newline").orElse(true)) {
```
Comment (jshook) [inline at-stdout/src/main/java/io/engineblock/activitytypes/stdout/StdoutActivity.java]: This block should not be removed, it's a different concern.

File: at-stdout/src/main/java/io/engineblock/activitytypes/stdout/StdoutActivity.java [hunk 3]
```diff
@@ -132,15 +132,16 @@ protected Writer createPrintWriter() {
...
-                    statement = statement+"\n";
-                }
+                Bindings bindings = bt.resolveBindings();
+                StringBindings sb = new StringBindings(compositor, bindings);
```
Comment (jshook) [inline at-stdout/src/main/java/io/engineblock/activitytypes/stdout/StdoutActivity.java]: This seems to be the same construct as used before, only via a different path. Specifically, the StringCompositor is still used, and its parseSection logic is the culprit. parseSection is called unconditionally from all constructors.

I have some upstream refactoring in the works that can simplify all of this by providing a single source of parsing logic. I'll reach out to review it with you in the am if you are available.

**Comment by jshook** [comment]
After looking at the code again, I realized we had a simple solution to start.
The compositor parser will honor escapes to any character by default, including '{'
The proper way to handle this for that version of the compositor would have been to simply add escapes to the non anchor curly braces.

However this highlighted another issue. The way that these are handled between stdout and other activities was not consistent. The change I'm putting in upstream will solve for this by putting common logic in virtdata for this type of tokenization, and also making the string compositor use it internally. 

The main difference here is that now the compositor will assume all parts of a raw template are not named anchors unless they are part of a specific pattern like {name} or ?name, where name can include dashes, dots, underscores, word characters, and numbers, but must start with a word character.
On the flip side, if we *needed* to have a template for stdout which used one of these patterns _as literal_, then we can't support that right now, as it would be matched as an anchor. We need to discuss these scenarios in a bit more detail, because I think they are prone to create other bugs. What we need to be able to do is modify the matching logic to support escapes as well without losing the simple form.

**Comment by jshook** [comment]
Since there is a fix with more consolidated upstream changes, I'm closing this PR.

**Pull request closed without merging**